Await approval request in transaction controller

matthewwalsh0 opened the pull request:
## Explanation

Refactor the transaction controller to decouple the public methods from the concept of confirmation by instead awaiting the approval request promise and then performing the approve or cancel logic, all in a single method execution.

At a technical level, this includes:

- Updating the `newUnapprovedTransaction` and `addUnapprovedTransaction` methods to await approval and continue with the post-approval logic.
- Changing the `approveTransaction` and `cancelTransaction` controller methods to private.
- Adding an optional `requiresApproval` property to `addUnapprovedTransaction` to skip approval.
- Updating the `updateAndApproveTx`, `cancelTx`, and `cancelTxs` actions to resolve or reject the approval using the `ApprovalController`.
- Updating the swaps reducer to add the transaction and skip approval in a single method call.
- Moving swaps specific transaction logic after creating the transaction but before automatic approval, into the transaction controller.

## Manual Testing Steps

Verify all transaction related flows including:

- Transfers
- Contract Operations
- Swaps

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

matthewwalsh0 committed:
Await approval request in transaction controller

Create temporary patch for approval controller.
Move swaps specific logic to background.
Resolve transaction approval request from frontend.

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Fix failing unit tests

matthewwalsh0 committed:
Handle persisted confirmations

Fix linting.

matthewwalsh0 committed:
Fix E2E tests

matthewwalsh0 committed:
Remove txMeta from approval request

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Create additional unit tests

matthewwalsh0 committed:
Fix bad method call

matthewwalsh0 committed:
Reject approval when cancelling multiple

Create unit tests for requireApproval option.

matthewwalsh0 committed:
Move swaps simulation error

Create associated unit tests.
Rename extraMeta property.

matthewwalsh0 committed:
Merge branch 'develop' into chore/462-await-transaction-approval-request

matthewwalsh0 committed:
Remove retry

matthewwalsh0 committed:
Create unit tests for signing and publish errors

matthewwalsh0 committed:
Update approval controller patch

Specify expectsResult in approval requests.

matthewwalsh0 committed:
Create additional unit tests

Rename _addTransaction to _createTransaction.

matthewwalsh0 committed:
Use new result property

matthewwalsh0 committed:
Create result callback unit tests

matthewwalsh0 committed:
Merge branch 'develop' into chore/462-await-transaction-approval-request

matthewwalsh0 committed:
Use released approval controller

matthewwalsh0 committed:
Remove Yarn duplicates

matthewwalsh0 committed:
Update LavaMoat policies

Taniel13 commented:
> ## Explanation
> 
> Refactor the transaction controller to decouple the public methods from the concept of confirmation by instead awaiting the approval request promise and then performing the approve or cancel logic, all in a single method execution.
> 
> At a technical level, this includes:
> 
> - Updating the `newUnapprovedTransaction` and `addUnapprovedTransaction` methods to await approval and continue with the post-approval logic.
> - Adding an optional `requiresApproval` property to `addUnapprovedTransaction` to skip approval.
> - Updating the `updateAndApproveTx`, `cancelTx`, and `cancelTxs` actions to resolve or reject the approval using the `ApprovalController`.
> - Updating the swaps reducer to add the transaction and skip approval in a single method call.
> - Moving swaps specific transaction logic after creating the transaction but before automatic approval, into the transaction controller.
> 
> ## Manual Testing Steps
> 
> Verify all transaction related flows including:
> 
> - Transfers
> - Contract Operations
> - Swaps
> 
> ## Pre-merge author checklist
> 
> - [x] I've clearly explained:
>   - [x] What problem this PR is solving
>   - [x] How this problem was solved
>   - [x] How reviewers can test my changes
> - [x] Sufficient automated test coverage has been added
> 
> ## Pre-merge reviewer checklist
> 
> - [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
> - [ ] PR is linked to the appropriate GitHub issue
> - [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone
> 
> If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.
> 
> In this case, a QA Engineer approval will be be required.
>

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1835,6 +1656,252 @@ export default class TransactionController extends EventEmitter {
+      let existingTxMeta =
+        this.txStateManager.getTransactionWithActionId(actionId);
+      if (existingTxMeta) {
+        this.emit('newUnapprovedTx', existingTxMeta);
```
Are we emitting this event only for the unit tests?

If so, we might want to think on how to refactor them, even if it's in a future PR.

matthewwalsh0 replied:
Didn't really like this as I wrote it. I've removed the event and am instead using a spy on the state manager.

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1835,6 +1656,252 @@ export default class TransactionController extends EventEmitter {
+    }
+
+    this.addTransaction(txMeta);
+    this.emit('newUnapprovedTx', txMeta);
```
Same comment as the previous `newUnapprovedTx` event, but this one is even weirder as `addTransaction` emits a ``${txMeta.id}:unapproved`` with exactly the same data and I cannot see if anyone is listening.

Odd, although we might want to keep them just in case, if they don't seem necessary I'd put a comment for future maintainers.

matthewwalsh0 replied:
I've left all the state manager events but removed `newUnapprovedTx` as it was now unused.

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1835,6 +1656,252 @@ export default class TransactionController extends EventEmitter {
+    this.emit('newUnapprovedTx', txMeta);
+
+    txMeta = await this.addTransactionGasDefaults(txMeta);
+
```
Totally optional and subjective, but I think it'd be clearer for anyone else reviewing if this whole block of code that comes from `swaps.js` UI logic was moved into a private method, guarded here by just the transaction type. Something like:

```
if (transactionType === TransactionType.swap) {
  this._whateverNameGivenToTheSwapSpecificWork(...)
}
```

But maybe this comment is good enough for any reviewer to see where it comes from and why it's here.

matthewwalsh0 replied:
Good idea, done.

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -798,99 +728,29 @@ export default class TransactionController extends EventEmitter {
+      options,
+    );
+    if (isExisting) {
+      return txMeta.status === TransactionStatus.unapproved
```
From what I can see, `_waitForTransactionFinished` resolves when a `${id}:finished` event is fired.

However, that event is only fired when the status is changed to `submitted`, `rejected` or `failed`.

Since we are just awaiting if the initial status is `unapproved`, is there any chance that we could be missing out on awaiting for the tx from any of the other states (`approved`, `signed`, `dropped`, `confirmed`, `pending`). It might be the case that this is impossible, in which case we are good, but just mentioning it so that it's not overlooked.

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -2662,56 +2729,70 @@ export default class TransactionController extends EventEmitter {
+        if (error.code === errorCodes.provider.userRejectedRequest) {
+          await this._cancelTransaction(txId, actionId);
+        } else {
+          await this._failTransaction(txId, error, actionId);
```
This one returns void.

matthewwalsh0 replied:
Done.

bergarces commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1361,10 +1205,8 @@ export default class TransactionController extends EventEmitter {
-   * @param opts.hasApprovalRequest - whether the transaction has an approval request
    */
-  async approveTransaction(txId, actionId, { hasApprovalRequest = true } = {}) {
+  async approveTransaction(txId, actionId) {
```
This is a weird one but... now that this is not called externally via `TransactionController` instance, but only via `PendingTransactionTracker` via passing the method in the constructor... should it be private?

matthewwalsh0 replied:
I didn't notice `PendingTransactionTracker` was constructed inside this controller, so changed to private to encapsulate everything.

cryptotavares commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -356,42 +356,44 @@ export default class TransactionController extends EventEmitter {
+    const txId = initialTxMeta.id;
+    const isCompleted = initialTxMeta.status !== TransactionStatus.unapproved;
+
+    const finishedPromise = isCompleted
+      ? Promise.resolve(initialTxMeta)
+      : this._waitForTransactionFinished(txId);
+
+    if (!isExisting && !isCompleted) {
+      try {
+        await this._requestTransactionApproval(initialTxMeta);
+      } catch (error) {
+        // Errors generated from final status using finished event
+      }
```
This is a farfetched case, but if the `_requestTransactionApproval` ends up throwing after setting the transaction status to approved but before setting it to submitted, then we will end up with a pending Promise (`finishedPromise`) that will never resolve and will be blocking the execution flow. Should we or even do we want to handle transactions that were not submitted but were a approved? 🤔 

(I guess that this could already happen before anyway, so any solution should be done on a subsequent PR)

matthewwalsh0 replied:
Good spot, I've fixed the condition inside the error handling in `_requestTransactionApproval` so it will now correctly cancel or fail the transaction if it's in any incomplete status. Meaning any errors here are just for information, and the transaction will always be `submitted`, `rejected`, or `failed` at this point.

matthewwalsh0 committed:
Make approveTransaction private

Add missing JSDoc.
Move swaps logic to separate private method.
Move unit test.

matthewwalsh0 committed:
Merge branch 'develop' into chore/462-await-transaction-approval-request

matthewwalsh0 committed:
Update Yarn lock

matthewwalsh0 committed:
Update LavaMoat policies

matthewwalsh0 commented on ui/store/actions.ts:
```diff
@@ -1154,18 +1103,22 @@ export function addUnapprovedTransactionAndRouteToConfirmationPage(
  * @param type - The type of the transaction being added.
+ * @param options
+ * @param options.requireApproval
+ * @param options.extraMeta
```
Updated.

matthewwalsh0 committed:
Cancel or fail on any incomplete status

danjm commented:
One issue I notice while reviewing this PR is we now have two different uses of the term "approval" within the transaction controller. This is not a huge deal, and it is a problem that existed before this PR, but the changes in this PR make it more noticeable.

To clarify, we have:
1. the concept of a "swapApproval", which is a transaction that calls the ERC20 `approve` method to grant the swaps contract the ablity to transfer tokens
2. the `ApprovalController` and the approvals it controllers

And so it can lead to some confusion for the reader of the code. For example we have code like this:
```
finalApproveTxMeta = await addUnapprovedTransaction(
          undefined,
          { ...approveTxParams, amount: '0x0' },
          TransactionType.swapApproval,
          {
            requireApproval: false,
```
Where `swapApproval` and `requireApproval` are right next to each other, but the `Approval`s in those two property names have entirely different meanings.

Now, this is not a huge deal, because when you read the transaction controller itself it is clear and well written. But a possible source of confusion. Including in `metamask-controller.js`, where we have the handling of the `newSwapApproval` event, but also lots of references to the `approvalController` and approvals that come from that controller, but no documentation to distinguish the `newSwapApproval` from the other sort of approvals.

We probably shouldn't address this problem in this PR, as it was a pre-existing problem. So I'd suggest a clean up task to follow this PR that renames the "swapApproval" to something else, perhaps "swapERC20Approve"

danjm commented on app/scripts/controllers/transactions/index.js:
```diff
@@ -1646,136 +1395,519 @@ export default class TransactionController extends EventEmitter {
+    }
+
+    if (transactionType === TransactionType.swapApproval) {
+      this.emit('newSwapApproval', txMeta);
```
What's the rationale for emitting an event here instead of just calling the swaps controller method here?

danjm replied:
Oh, maybe because the transactions controller has to be instantiated before the swaps controller?

matthewwalsh0 replied:
Exactly, a circular dependency. We would need the `SwapsController` to exist first to provide the callback in the constructor, but the `SwapsController` requires the `TransactionController` for callbacks in its constructor.

An event enables us to create the listener after both are already constructed.

danjm commented on app/scripts/background.js:
```diff
@@ -676,10 +676,8 @@ export function setupController(
   //
   // User Interface setup
   //
```
Why do we no longer need to wait for `initApprovals()` to call `updateBadge()`?

matthewwalsh0 replied:
The previous implementation of `initApprovals` wouldn't resolve until the requests were approved, meaning the badge number would be zero at that point anyway.

Plus I noticed we already listened to the `updateBadge` event from the approval controller below anyway, meaning the badge is already reacting to the creation of the approval requests.

danjm commented on ui/store/actions.ts:
```diff
@@ -1663,15 +1623,21 @@ export function cancelTxs(
-              }
-
-              resolve();
-            });
```
It took me a little while to figure out how the transaction state would correctly be updated now that we are not calling `cancelTransaction` directly. But I think I get it now:
1. `rejectPendingApproval` is called, with the second parameter being the error created by `ethErrors.provider.userRejectedRequest().serialize()`
2. `this.approvalController.reject` is called with an error constructed from its second parameter
3. Steps 1 and 2 will always be happening while we are awaiting `this._requestApproval` in `_requestTransactionApproval` the transaction controller
4. If the user rejects the request, `this._requestApproval` will throw an error
5. The error handler in `_requestTransactionApproval` will check that the error code matches what was set in step 2, and then call `this._cancelTransaction`

Does that sound right?

danjm replied:
Also, I notice that we are passing `ethErrors.provider.userRejectedRequest().serialize()` here, but in `rejectPendingApproval` in `metamask-controller.js` we assume that the error parameter is not serialized (`new EthereumRpcError(error.code, error.message, error.data)`).

Does deserialization happen automatically by some part of our ui<>background communication code?

matthewwalsh0 replied:
Exactly right.

The `serialize` method is a bit misleading as it's not serialising to a string, but rather a simple object with only the public properties remaining, including `code`.

https://github.com/MetaMask/rpc-errors/blob/v4.0.3/src/classes.ts#L44

danjm reviewed (changes requested):
This looks good!

I have left a few questions and when they are answered I can approve.

danjm reviewed (dismissed):
Nice work!

matthewwalsh0 committed:
Merge remote-tracking branch 'origin/develop' into chore/462-await-transaction-approval-request
